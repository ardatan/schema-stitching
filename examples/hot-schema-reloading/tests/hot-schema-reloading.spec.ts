import { createServer } from 'http';
import { createSchema, createYoga } from 'graphql-yoga';
import { startServer, stopServer } from '../src/gateway';
import { inventoryServer } from '../src/services/inventory/server';
import { productsServer } from '../src/services/products/server';

describe('Hot schema reloading', () => {
  let cnt = 0;
  const fooServer = createServer(
    createYoga({
      schema: () => {
        const typeDefs = /* GraphQL */ `
                    type Query {
                        """
                        Count: ${cnt++}
                        """
                        foo: String!
                        _sdl: String!
                    }
                `;
        return createSchema<any>({
          typeDefs,
          resolvers: {
            Query: {
              foo: () => 'bar',
              _sdl: () => typeDefs,
            },
          },
        });
      },
    }),
  );
  beforeEach(async () => {
    await Promise.all([
      startServer(),
      new Promise<void>(resolve => productsServer.listen(4001, resolve)),
      new Promise<void>(resolve => inventoryServer.listen(4002, resolve)),
      new Promise<void>(resolve => fooServer.listen(4003, resolve)),
    ]);
  });
  afterEach(async () => {
    cnt = 0;
    await Promise.all([
      stopServer(),
      new Promise(resolve => productsServer.close(resolve)),
      new Promise(resolve => inventoryServer.close(resolve)),
      new Promise(resolve => fooServer.close(resolve)),
    ]);
  });
  async function fetchEndpoints() {
    const response = await fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          query {
            endpoints {
              url
              sdl
            }
          }
        `,
      }),
    });
    const result = await response.json();
    return result;
  }
  it('should add/remove endpoints correctly', async () => {
    const removeResponse = await fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          mutation {
            removeEndpoint(url: "http://localhost:4002/graphql") {
              success
            }
          }
        `,
      }),
    });
    const removeResult = await removeResponse.json();
    expect(removeResult).toMatchSnapshot('removeResult');
    const removedEndpointsResult = await fetchEndpoints();
    expect(removedEndpointsResult).toMatchSnapshot('removedEndpointsResult');
    const addResponse = await fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          mutation {
            addEndpoint(url: "http://localhost:4002/graphql") {
              success
              endpoint {
                url
                sdl
              }
            }
          }
        `,
      }),
    });
    const addResult = await addResponse.json();
    expect(addResult).toMatchSnapshot('addResult');
    const addedEndpointsResult = await fetchEndpoints();
    expect(addedEndpointsResult).toMatchSnapshot('addedEndpointsResult');
  });
  it('should reload schemas correctly', async () => {
    const addResponse = await fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          mutation {
            addEndpoint(url: "http://localhost:4003/graphql") {
              success
              endpoint {
                url
                sdl
              }
            }
          }
        `,
      }),
    });
    const addResult = await addResponse.json();
    expect(addResult).toMatchSnapshot('addResult');
    const reloadEndpointsResponse = await fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          mutation {
            reloadAllEndpoints {
              success
            }
          }
        `,
      }),
    });
    const reloadEndpointsResult = await reloadEndpointsResponse.json();
    expect(reloadEndpointsResult).toMatchSnapshot('reloadEndpointsResult');
    const endpointsResult = await fetchEndpoints();
    expect(endpointsResult).toMatchSnapshot('endpointsResult');
  });
});
