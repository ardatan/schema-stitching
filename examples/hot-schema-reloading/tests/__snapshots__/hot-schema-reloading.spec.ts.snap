// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Hot schema reloading should add/remove endpoints correctly: addResult 1`] = `
{
  "data": {
    "addEndpoint": {
      "endpoint": {
        "sdl": "
  directive @key(selectionSet: String!) on OBJECT
directive @computed(selectionSet: String!) on FIELD_DEFINITION
directive @merge(argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) on FIELD_DEFINITION
directive @canonical on OBJECT | INTERFACE | INPUT_OBJECT | UNION | ENUM | SCALAR | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  type Product {
    upc: ID!
    inStock: Boolean
  }

  type Query {
    mostStockedProduct: Product
    _products(upcs: [ID!]!): [Product]! @merge(keyField: "upc")
    _sdl: String!
  }
",
        "url": "http://localhost:4002/graphql",
      },
      "success": true,
    },
  },
}
`;

exports[`Hot schema reloading should add/remove endpoints correctly: addedEndpointsResult 1`] = `
{
  "data": {
    "endpoints": [
      {
        "sdl": "
  directive @key(selectionSet: String!) on OBJECT
directive @computed(selectionSet: String!) on FIELD_DEFINITION
directive @merge(argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) on FIELD_DEFINITION
directive @canonical on OBJECT | INTERFACE | INPUT_OBJECT | UNION | ENUM | SCALAR | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  type Product {
    upc: ID!
    inStock: Boolean
  }

  type Query {
    mostStockedProduct: Product
    _products(upcs: [ID!]!): [Product]! @merge(keyField: "upc")
    _sdl: String!
  }
",
        "url": "http://localhost:4002/graphql",
      },
      {
        "sdl": "
  directive @key(selectionSet: String!) on OBJECT
directive @computed(selectionSet: String!) on FIELD_DEFINITION
directive @merge(argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) on FIELD_DEFINITION
directive @canonical on OBJECT | INTERFACE | INPUT_OBJECT | UNION | ENUM | SCALAR | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  type Product {
    upc: ID!
    name: String!
    price: Int!
    weight: Int!
  }

  type Query {
    topProducts(first: Int = 2): [Product]!
    products(upcs: [ID!]!): [Product]! @merge(keyField: "upc")
    _sdl: String!
  }
",
        "url": "http://localhost:4001/graphql",
      },
    ],
  },
}
`;

exports[`Hot schema reloading should add/remove endpoints correctly: removeResult 1`] = `
{
  "data": {
    "removeEndpoint": {
      "success": true,
    },
  },
}
`;

exports[`Hot schema reloading should add/remove endpoints correctly: removedEndpointsResult 1`] = `
{
  "data": {
    "endpoints": [
      {
        "sdl": "
  directive @key(selectionSet: String!) on OBJECT
directive @computed(selectionSet: String!) on FIELD_DEFINITION
directive @merge(argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) on FIELD_DEFINITION
directive @canonical on OBJECT | INTERFACE | INPUT_OBJECT | UNION | ENUM | SCALAR | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  type Product {
    upc: ID!
    name: String!
    price: Int!
    weight: Int!
  }

  type Query {
    topProducts(first: Int = 2): [Product]!
    products(upcs: [ID!]!): [Product]! @merge(keyField: "upc")
    _sdl: String!
  }
",
        "url": "http://localhost:4001/graphql",
      },
    ],
  },
}
`;

exports[`Hot schema reloading should reload schemas correctly: addResult 1`] = `
{
  "data": {
    "addEndpoint": {
      "endpoint": {
        "sdl": "
                    type Query {
                        """
                        Count: 0
                        """
                        foo: String!
                        _sdl: String!
                    }
                ",
        "url": "http://localhost:4003/graphql",
      },
      "success": true,
    },
  },
}
`;

exports[`Hot schema reloading should reload schemas correctly: endpointsResult 1`] = `
{
  "data": {
    "endpoints": [
      {
        "sdl": "
  directive @key(selectionSet: String!) on OBJECT
directive @computed(selectionSet: String!) on FIELD_DEFINITION
directive @merge(argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) on FIELD_DEFINITION
directive @canonical on OBJECT | INTERFACE | INPUT_OBJECT | UNION | ENUM | SCALAR | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  type Product {
    upc: ID!
    name: String!
    price: Int!
    weight: Int!
  }

  type Query {
    topProducts(first: Int = 2): [Product]!
    products(upcs: [ID!]!): [Product]! @merge(keyField: "upc")
    _sdl: String!
  }
",
        "url": "http://localhost:4001/graphql",
      },
      {
        "sdl": "
                    type Query {
                        """
                        Count: 1
                        """
                        foo: String!
                        _sdl: String!
                    }
                ",
        "url": "http://localhost:4003/graphql",
      },
      {
        "sdl": "
  directive @key(selectionSet: String!) on OBJECT
directive @computed(selectionSet: String!) on FIELD_DEFINITION
directive @merge(argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) on FIELD_DEFINITION
directive @canonical on OBJECT | INTERFACE | INPUT_OBJECT | UNION | ENUM | SCALAR | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  type Product {
    upc: ID!
    inStock: Boolean
  }

  type Query {
    mostStockedProduct: Product
    _products(upcs: [ID!]!): [Product]! @merge(keyField: "upc")
    _sdl: String!
  }
",
        "url": "http://localhost:4002/graphql",
      },
    ],
  },
}
`;

exports[`Hot schema reloading should reload schemas correctly: reloadEndpointsResult 1`] = `
{
  "data": {
    "reloadAllEndpoints": {
      "success": true,
    },
  },
}
`;
